onnxruntime/test/providers/cpu/nn/group_norm_op_test.cc

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "gtest/gtest.h"
+#include "test/providers/provider_test_utils.h"
+#include "test/common/tensor_op_test_utils.h"
+#include "test/util/include/default_providers.h"
+
+#ifdef COREML_ENABLE_MLPROGRAM
+using namespace std;
+namespace onnxruntime {
+namespace test {
+
+template <typename T>
+class GroupNormalizationOpTest : public ::testing::Test {
+};
+using GroupNormalizationOpTestTypes = ::testing::Types<float, MLFloat16>;
+TYPED_TEST_SUITE(GroupNormalizationOpTest, GroupNormalizationOpTestTypes);
+
+template <typename T>
+static std::vector<T> GetTypedArray(std::vector<float> inputs, [[maybe_unused]] T v = T(0.f)) {
+  if constexpr (std::is_same<T, float>::value) {
+    return inputs;
+  } else {
+    return ToFloat16(inputs);
+  }
+}
+
+// Disable TensorRT on some of the tests because its parser doesn't support weight as input
+
+TYPED_TEST(GroupNormalizationOpTest, InstanceNorm) {
+  OpTester test("GroupNormalization", 18);
+  test.AddAttribute("epsilon", 0.3F);
+  test.AddAttribute("num_groups", int64_t(3));
+
+  vector<float> input = {3.1513367F, 9.283596F, 1.4546119F, 5.4617004F,
+                         8.519701F, 1.2382338F, 1.7930176F, 5.1099434F,
+                         7.9195533F, 7.638727F, 8.065445F, 3.8082376F,
+
+                         2.3667817F, 2.8248506F, 3.7754705F, 5.861325F,
+                         5.058735F, 3.2787242F, 3.6843839F, 9.755121F,
+                         2.7902672F, 7.3974323F, 8.283609F, 8.488337F};
+  vector<int64_t> input_dims = {2, 3, 4};
+  test.AddInput<TypeParam>("X", input_dims, GetTypedArray<TypeParam>(input));
+
+  vector<float> scale = {1.F, 1.F, 1.F};
+  vector<int64_t> scale_dims = {3};
+  test.AddInput<TypeParam>("scale", scale_dims, GetTypedArray<TypeParam>(scale), true);
+
+  vector<float> B = {0.F, 0.F, 0.F};
+  vector<int64_t> B_dims = {3};
+  test.AddInput<TypeParam>("bias", B_dims, GetTypedArray<TypeParam>(B), true);
+
+  vector<float> expected_output = {-0.56495477f, 1.48930046f, -1.13334329f, 0.20899761f,
+                                   1.46688162f, -0.98600774f, -0.79911913f, 0.31824524f,
+                                   0.57370438f, 0.42193634f, 0.6525492f, -1.64818992f,
+
+                                   -0.92380346f, -0.60808484f, 0.04711878f, 1.48476953f,
+                                   -0.14644464f, -0.82262872f, -0.66852817f, 1.63760153f,
+                                   -1.65898662f, 0.27618144f, 0.64840618f, 0.734399f};
```

Comment:
How was the expected output calculated? Just want to be extra careful it's correct given we're composing the coreml implementation.

Reply:
It's calcucatd by torch.nn.group_norm

Reply:
Would be great to add a comment saying that.

Reply:
Sure.